annif/backend/yake.py

```diff
@@ -0,0 +1,191 @@
+    }
+
+    def default_params(self):
+        params = backend.AnnifBackend.DEFAULT_PARAMETERS.copy()
+        params.update(self.DEFAULT_PARAMETERS)
+        return params
+
+    @property
+    def is_trained(self):
+        return True
+
+    @property
+    def label_types(self):
+        if type(self.params['label_types']) == str:  # Label types set by user
+            label_types = [lt.strip() for lt
+                           in self.params['label_types'].split(',')]
+            self._validate_label_types(label_types)
+        else:
+            label_types = self.params['label_types']  # The defaults
+        return [getattr(SKOS, lt) for lt in label_types]
+
+    def _validate_label_types(self, label_types):
+        for lt in label_types:
+            if lt not in ('prefLabel', 'altLabel', 'hiddenLabel'):
+                raise ConfigurationException(
+                    f'invalid label type {lt}', backend_id=self.backend_id)
+
+    def initialize(self):
+        self._initialize_index()
+
+    def _initialize_index(self):
+        if self._index is None:
+            path = os.path.join(self.datadir, self.INDEX_FILE)
+            if os.path.exists(path):
+                self._index = self._load_index(path)
+                self.debug(
+                    f'Loaded index from {path} with {len(self._index)} labels')
+            else:
+                self.info('Creating index')
+                self._index = self._create_index()
+                self._save_index(path)
+                self.info(f'Created index with {len(self._index)} labels')
+
+    def _save_index(self, path):
+        with open(path, 'w', encoding='utf-8') as indexfile:
+            for label, uris in self._index.items():
+                line = label + '\t' + ' '.join(uris)
+                print(line, file=indexfile)
+
+    def _load_index(self, path):
+        index = dict()
+        with open(path, 'r', encoding='utf-8') as indexfile:
+            for line in indexfile:
+                label, uris = line.strip().split('\t')
+                index[label] = uris.split()
+        return index
+
+    def _create_index(self):
+        index = defaultdict(set)
+        skos_vocab = self.project.vocab.skos
+        for concept in skos_vocab.concepts:
+            uri = str(concept)
+            labels = skos_vocab.get_concept_labels(
+                concept, self.label_types, self.params['language'])
+            for label in labels:
+                label = self._normalize_label(label)
+                index[label].add(uri)
+        index.pop('', None)  # Remove possible empty string entry
+        return dict(index)
+
+    def _normalize_label(self, label):
+        label = str(label)
+        if annif.util.boolean(self.params['remove_parentheses']):
+            label = re.sub(r' \(.*\)', '', label)
+        lemmatized_label = self._lemmatize_phrase(label)
+        return self._sort_phrase(lemmatized_label)
+
+    def _lemmatize_phrase(self, phrase):
```

Comment:
I think this could be renamed to `_normalize_phrase` because not all analyzers perform lemmatization (only Voikko does currently). Snowball does stemming but that's a bit different...